platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java

```diff
@@ -28,31 +28,15 @@
 public interface ProblemSpec {
 
     /**
-     * Defines simple identification for this problem.
+     * Defines the context-independent identifier for this problem.
      * <p>
-     * It is a mandatory property to configure when emitting a problem with {@link ProblemReporter}..
-     * <p>
-     * Calling this method will set the reported problem group to {@link SharedProblemGroup#generic()}
+     * It is a mandatory property to configure when emitting a problem with {@link ProblemReporter} and can be defined by using {@link IdFactory}.
      *
-     * @param name the name of the problem. As a convention kebab-case-formatting should be used.
-     * @param displayName a human-readable representation of the problem, free of any contextual information.
+     * @param problemId the problem id
      * @return this
-     * @since 8.8
-     */
-    ProblemSpec id(String name, String displayName);
-
-    /**
-     * Defines simple identification for this problem.
-     * <p>
-     * It is a mandatory property to configure when emitting a problem with {@link ProblemReporter}.
-     *
-     * @param name the name of the problem. As a convention kebab-case-formatting should be used.
-     * @param displayName a human-readable representation of the problem, free of any contextual information.
-     * @param parent the container problem group.
-     * @return this
-     * @since 8.8
+     * @since 8.13
      */
-    ProblemSpec id(String name, String displayName, ProblemGroup parent);
+    ProblemSpec id(ProblemId problemId);
```

Comment:
:x: I think this is a big usability problem: how will the users figure out that they need to use `IdFactory` to get an ID? And I don't think a JavaDoc is a good answer here, as they will first look at the `ProblemId`. 

I think we might need to rethink this idea and introduce the factory into the `ProblemId` itself so we can provide a reading-less, more natural way of creating IDs.

Reply:
I don't think that javadoc is all that bad. At the same time, I'm also not happy with the disconnectedness of `IdFactory`.

Right now the most sensible change would be to turn ProblemID and ProblemGroup to abstract classes and the static factories there:

```java
public abstract class ProblemId {

    public abstract String getName();
    public abstract String getDisplayName();
    public abstract ProblemGroup getGroup();

    public static ProblemId create(String name, String displayName, ProblemGroup group) {
        return new DefaultProblemId(name, displayName, group);
    }
}
```

(similar for ProblemGroup).